### PR TITLE
[wip] crossword title slugs

### PR DESCRIPTION
**Pull request by ckolderup**
I think this is a start on #86, but there's probably some work left to go. One thing that occurred to me is that it'd be nice if there were a redirect when you go to `/crossword/{id}` to a new canonical `/crossword/{id}/slug`, but I don't know off the top of my head how to do that in next.js (you could do it manually with history.push on mount for that page, I guess? But maybe it's easier than that?)

I also haven't checked the URLs in a couple of the places that I changed them being generated-- I need to make myself an admin in my dev environment so I can see some of the pages.

### CODE REVIEW

File: app/components/ConstructorStats.tsx [hunk 1]
```diff
@@ -59,7 +59,7 @@ const PuzzleLink = ({
   }
   return (
     <li>
-      <Link href={'/crosswords/' + puzzleId}>
+      <Link href={`/crosswords/${puzzleId}/${slugify(puzzle?.t)}`}>
```
Comment (mdirolf) [inline app/components/ConstructorStats.tsx:62]: Does `slugify` handle nulls/undefineds? I don't think it does so I'm surprised this isn't a type error. If I'm right then we should just add an `|| ''` here so the link will just have an empty slug param until puzzle exists

File: app/pages/api/puz/[[...puzzleId]].ts [hunk 2]
```diff
@@ -3,11 +3,15 @@ import { exportFile } from '../../../lib/converter';
 import { getPuzzle } from '../../../lib/puzzleCache';
 
 export default async function puz(req: NextApiRequest, res: NextApiResponse) {
-  const { puzzleId } = req.query;
-  if (!puzzleId || Array.isArray(puzzleId)) {
+  const { urlTokens } = req.query;
+  const [ puzzleId ] = urlTokens || '';
+  console.log('here we are');
```
Comment (mdirolf) [inline app/pages/api/puz/[[...puzzleId]].ts:8]: Just a note to remove this extra logging

Comment (mdirolf) [reply app/pages/api/puz/[[...puzzleId]].ts:8]: Actually - I think maybe we should just leave this endpoint completely unchanged? Maybe this was an accidental commit. But this is what generates links to dl .puz files and I don't think we need to change those at all

**Comment by mdirolf** [comment]
Overall this looks great! 👏👏 You found all the places I could think of where this needed to change, and some I didn't think of 😂.

There is a script `app/scripts/makeAdmin.ts` which I think should be able to make you an admin user locally. YMMV though I haven't used it in almost 2 years so it might need some updating.

I do think you're right that we need to handle redirects better - my biggest worry w/ this change is the SEO implications of changing so many URLs at once. I think the correct way to do it is to `301` from getServerSideProps - check out the redirect response here: https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering

We should make sure that the following urls all redirect correctly:
/crosswords/XXX (current url)
/crosswords/XXX/ (trailing slash)
/crosswords/XXX/incorrect-or-stale-slug (if a user changes title or a url gets clipped)

The only other thing I can think of is that I think we should length limit the urls somehow - probably by adding a limit to `slugify`. I believe the only other place that gets used is for article urls so it'd be fine to have a limit there as well. No idea what best practice is but certainly something like 150 chars in the slug ought to be enough for any reasonable title while still keeping urls manageable

**Comment by mdirolf** [comment]
I also just thought of one other hiccup - these two pages:
/crosswords/XXX/edit to edit one of your puzzles
/crosswords/XXX/stats to view puzzle stats

If somebody happened to name their puzzle "Stats" or "EDIT!" or something then those pages would take priority over the catch-all and the puzzle links would all be dead. The options I can think of are:
- Move those two pages somewhere else, like /edit/XXX or /stats/XXX
- Ensure that slugify never returns "edit" or "stats", instead munging it into "edit-the-puzzle" or something

Probably the former is more correct, I think? Doubt there are many links to those pages around that we need to worry about breaking, since they're both private pages. Only thing it'd break is if constructors have gotten used to manually changing the url and adding '/edit' or something, but 🤷

**Comment by mdirolf** [comment]
Thanks for getting the ball rolling here @ckolderup! I didn't want to leave this lingering so I made the changes I had requested and merged - should go live the next time the site is deployed 😊

**Comment by ckolderup** [comment]
Ah geez, thanks for seeing this one through. Sorry for letting it sit-- things got unexpectedly busy at work and home. Maybe next time I can bite off something a bit smaller-scoped 😅

**Comment by mdirolf** [comment]
No need to apologize at all! It's awesome to have more folks contributing to the project.

**Pull request merged**